design/design-doc-smaller-providers.md

```diff
@@ -0,0 +1,704 @@
+for some time. The Crossplane maintainers are now aligned that it’s necessary.
+We have explored the following options:
+
+* Make it possible to filter unnecessary CRDs by configuring a provider.
+* Break the biggest providers up into smaller, service-scoped providers.
+* Load CRDs into the API server lazily, on-demand.
+
+## Goals
+
+The goal of this proposal is to eliminate _or significantly defer_ the
+performance issues that arise when installing too many CRDs. By "deferring"
+performance issues, I mean buying time for the Kubernetes ecosystem to improve
+its performance under CRD load.
+
+## Non-goals
+
+The goal is not that no Crossplane deployment _ever_ crosses the ~500 CRD mark.
+Instead, that should be a rare edge case until the vast majority of Kubernetes
+deployments can handle it without issue.
+
+This proposal _does not_ intend to address any security concerns related to
+unused CRDs being present in a Crossplane deployment. At this time the
+significance of these security concerns is not well understood. It's also
+unclear how many community members are concerned, relative to the overwhelming
+concern about performance.
+
+At the present time I do not believe installing a CRD (and controller) that you
+don't intend to use poses a significant risk. At least two layers of security
+prevent anyone using the Crossplane Managed Resources (MRs) these CRDs define -
+Kubernetes RBAC in front of the API call, and cloud IAM permissions behind it.
+
+## Proposal
+
+I propose that provider maintainers break the largest providers into smaller,
+service-scoped providers. Each service-scoped provider would correspond to what
+is currently an API group within a provider. For example,
+`upbound/provider-aws-rds` would extend Crossplane with support for the 22
+managed resources (and thus 22 CRDs) in the `rds.aws.upbound.io` API group.
+
+### Overview
+
+I proposed we only break up the largest providers. For example:
+
+* `upbound/provider-aws` - Becomes ~150 smaller providers.
+* `upbound/provider-azure` - Becomes ~100 smaller providers.
+* `upbound/provider-gcp` - Becomes ~75 smaller providers.
+* `crossplane-contrib/provider-tencentcloud` - Becomes ~50 smaller providers.
+
+It's important to note that we don't expect anyone to be installing and
+operating all of these providers. Per [Community Impact](#community-impact), we
+expect folks to install ~10 providers on average.
+
+These providers would:
+
+* Be built from “monorepos” - the same repositories they are today, like
+  `upbound/provider-aws`.
+* Share a `ProviderConfig` by taking a dependency on a (e.g.)
+  `provider-aws-config` package.
+* Continue to use contemporary “compiled-in” cross-resource-references, short to
+  medium term.
+
+Provider generation tooling like [`upjet`][upjet] would be updated to support
+generating service-scoped providers.
+
+No changes would be required to Crossplane core - i.e. to the package manager.
```

Comment:
I need to update this. We did find in testing that we needed a small change to the RBAC manager, which you can see in https://github.com/crossplane/crossplane/pull/4010.

Reply:
Done.